### ISSUE TITLE
Loading large tables of data via python API bloats Inventree server memory usage

### ISSUE THREAD

**Issue by zoopsel** [inventree/InvenTree#7429]
### Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find a similar issue

### Describe the bug*

When running a python program that requests data via the inventree-python api from a large table, the inventree server's memory usage bloats in size to multiple gigabytes.
In can happen that during this time all the memory of the system is used up and linux kills the server.

We have quite a lot of StockItemTracking entries in our Inventree database (9.5 MB), and I would like to fetch them all ( `StockItemTracking.list(api)` ) in order to do some quick in memory processing on the data.
First, we had to increase the timeout value of the connection, as this is set to 10 seconds by default.
Then we are able to fetch all these elements from the server but it takes quite lot of time (>10mins) and the memory used by the Inventree process goes from 260MB to 8.2GB.
When the fetching process eventually finishes, the memory used by Inventree remains at around 4GB and does not decrease back to 260MB.

This happens when using the Inventree develop server (`invoke server`) and when using Inventree with Gunicorn.

Is that normal behaviour or might this be some caching bug?

### Steps to Reproduce

1. Have a table with lots of entries in the Inventree database such that the database table has a sixe in the order of MBs.
2. Create a python script and fetch this table via the api. (you probably will have to increase the timeout value)
3. Use a process monitor and look at memory usage.

### Expected behaviour

No memory bloat.

### Deployment Method

- [ ] Docker
- [ ] Package
- [X] Bare metal
- [ ] Other - added info in Steps to Reproduce

### Version Information

0.15.3 (but also on 0.12.9)

### Please verify if you can reproduce this bug on the demo site.

- [ ] I can reproduce this bug on the demo site.

### Relevant log output

```shell
Neither the server console nor the console of the python script fetching the data produce any output.
```

**Comment by SchrodingersGat** [comment]
Is this specific to fetching via the python bindings? Or any such request to the API?

In any case this is a serious issue, the memory should definitely not be bloating out like that.

- How many items are in the table you are calling?
- Does this happen with multiple tables or is limited to a single table?
- Any other testing you have performed which might inform this ?

**Comment by zoopsel** [comment]
I encountered the bug when using the python bindings, never tried other ways.
Other tables work fine. For example the StockItems i can fetch via python api bindings without any problems.
In our database there are 29'074 stock items and the whole table has a size of 4.5 MB.
In the stock item tracking table there are 103'669 items and the size is 9.5MB.

I quickly tried to use [memray](https://github.com/bloomberg/memray), a python memory profiler, by calling `memray3.10 run manage.py runserver 127.0.0.1:12997` but there I couldn't see the same increase in memory usage as when using `htop`.

I will check if the same bug happens when fetching the stock item tracking table without the python bindings.

**Comment by zoopsel** [comment]
Same memory bloat happens when directly calling `http://127.0.0.1:12997/api/stock/track/`.

**Comment by zoopsel** [comment]
I tried your suggested approach with manual paginating 1000 StockItemTracking items at a time. This works for me at the moment.
However, there is still an increase memory consumption when the first api call is made from around 260MB up to 450MB.
This higher memory usage is persistent until the server is restarted.

As for me, this issue can be closed but it might still be a bug / side effect that should be kept an eye on.

**Comment by SchrodingersGat** [comment]
This definitely deserves further attention. I will leave this open.

@zoopsel as no-one else has reported this, if you have any further metrics or test data from your setup it would be greatly appreciated. I am not suggesting that it is your setup causing this, but any info you can provide will help us debug / fix the memory bloats.

**Comment by SchrodingersGat** [comment]
@zoopsel there is a significant performance bug in the stock item tracking API 

Fix incoming here - https://github.com/inventree/InvenTree/pull/7451

Still trying to reproduce the memory bloat

**Comment by zoopsel** [comment]
Hi @SchrodingersGat 
Thanks for the update! The API query speedup is much appreciated!
I tested again with inventree 0.15.4 and cannot measure the excessive memory bloat anymore, even when fetching all 100k stock item tracking entries at once.
Seems like my bug was resolved with the improvements you made in #7451.

**Comment by SchrodingersGat** [comment]
@zoopsel thanks for reporting back, glad to hear that it is working better for you. Happy to close this issue out?
